### PR TITLE
New hire form flow #775

### PR DESCRIPTION
**Pull request by MikeRayUX**
This PR adds Flow for businesses to notify RailsDevs when they make a hire #764

- [x] Adds a "Made a hire?" link to the nav bar
- [x] Clicking the link shows a form that looks like /hired/forms/new

### CODE REVIEW

**Comment by MikeRayUX** [comment]
This PR contains prev commit progress from #775 and serves as my starting point.

**Comment by MikeRayUX** [comment]
Right now: 

'hire' = `business_forms`  (Business hired a developer)
'hired' = `developer_forms` (Developer found a job)

Both hire/hired follow their own namespace.

Are you satisfied with just the naming change to i18n keys/admin menu buttons? Or would you also like to update the namespaces?

**Comment by joemasilotti** [comment]
> Would you also like to update the namespaces as well?

That would be great, thank you!

**Comment by MikeRayUX** [comment]
> That would be great, thank you! Same goes for the routes, too.

Great. On it!

**Comment by MikeRayUX** [comment]
Hey @joemasilotti, I'm currently exploring potential options for changing the namespace from `:hire` to `:business_forms`.

Starting with the current `:hire` namespace:

```ruby
namespace :hire do
  resources :forms, only: [:index, :show]
end
```

(Option: 1) We could move take the `:hire` (`business_forms`) namespace and move it into `:businesses`

```ruby
resources :businesses, only: [] do
  #resources :conversations, only: :index, controller: :business_conversations
  #resources :invisiblizes, only: :create, module: :businesses
  resources :forms, only: [:index, :show]
end
```

(Option: 2) Rename the `:hire` namespace to `business_forms`:

```ruby
namespace :business_forms do
  resources :forms, only: [:index, :show] 🚿 
end

```

Any thoughts?

**Comment by joemasilotti** [comment]
URL-wise, I like `/businesses/forms` and `/developers/forms`. So whatever routes can make that works feel right to me.

**Comment by MikeRayUX** [comment]
Done.

`admin/hire/forms` is now `admin/businesses/forms`
and
`admin/hired/forms` is now `admin/developers/forms`

Any specific behavior would you like to see tests for? 

Admin views newly created business/developer form for instance?

**Comment by joemasilotti** [comment]
I recently added tests around the admin CRUD for specialties. Something like that for dev + biz would be perfect!

**Comment by MikeRayUX** [comment]
> I recently added tests around the admin CRUD for specialties. Something like that for dev + biz would be perfect!

Sounds good! On it.

**Comment by joemasilotti** [comment]
Amazing thank you! One more request if you’re down for it - I’d love to see a db seed for each form. I’m planning to review this tomorrow - thanks for picking it up!

**Comment by MikeRayUX** [comment]
I added a seed to `hiring.rb` in e545ddf1bcc8ea51aa7d6dce68aea65743273f76

Bit tricky to see because at the model level, the business_form still exists as `Hire:Form`.

https://github.com/joemasilotti/railsdevs.com/blob/486d887c4b0a2bce386548d9cad69099ca8302ef/db/seeds/hiring.rb#L14-L23

The model/migration would be the last remaining piece to change.

I think I can get that done before you review it tomorrow.

**Comment by MikeRayUX** [comment]
`Businesses::BusinessForm` is a bit redundant, but it prevents us from a potential table conflict from

* `Developesr::Form` 
* `Businesses::Form`.

Alternatively, we could create a generic `Form` model that handles both biz/dev forms with an enum and backfill existing `hired_forms`. This would also require changing some associations. Thinking that might be a good fit for a future issue.

**Comment by MikeRayUX** [comment]
The model `Hire::Form` is now `Businesses::BusinessForm`

Also, the migration from this PR that creates `hire_forms` will create `business_forms` instead.

`Businesses::BusinessForm` is a bit redundant but it prevents us from breaking out of the current table naming scheme and from doing something more complex like setting custom table_names on new models like:

* `Developers::Form`
* `Businesses::Form`

Which AFAIK would require creating a generic `Form` model that handles both biz/dev forms via an enum and backfilling existing `hired_forms`. This would also require changing some associations. RN I'm thinking that might be a good fit for a future issue.

I'm open to any/all feedback. Happy reviewing!

**Comment by MikeRayUX** [comment]
Ah! This looks promising!

So this means we could achieve `Businesses::Form` that uses a table called `businesses_forms`

Happy to give it a go today! Sound good?

**Comment by joemasilotti** [comment]
Sounds great - go for it!

**Comment by MikeRayUX** [comment]
In the future, I think it would be feasible to move existing `Hired::Form` `HiringAgreements::Term` `HiringAgreements::Signature` into `Forms::` namespace

Something like

`Hired::Form` -> `Forms::Developers::Hired`

`Business::BusinessForm` -> `Forms::Businesses::Hired` 

`HiringAgreements::Term` -> `Forms::HiringAgreements::Term`

`HiringAgreements::Signature` -> `Forms::HiringAgreements::Signature`

In this PR, I think we have an opportunity to kick that off with `Businesses::BusinessForm` (since the migration is brand new)

Starting with creating the prefix `forms_`

```ruby
# app/models/forms.rb

module Forms
  def self.table_name_prefix
    "forms_"
  end
end
```

Then the table:

```ruby
create_table :forms_businesses_hired do |t|
   ...
end
```

Then we can change: `Business::BusinessForm` -> `Forms::Businesses::Hired` 

What do you think?

**Comment by joemasilotti** [comment]
I love this! Great idea.

So this PR would only make changes to the new model and you would open a new issue/PR for the remaining refactor?

**Comment by joemasilotti** [comment]
Sounds good all around to me. Let's move forward without the refactor for now and get this PR merged.

Is everything we talked about implemented? Or do you still need to add the `Forms` module?

**Comment by MikeRayUX** [comment]
Sounds good! Still working on Business::BusinessForm -> Forms::Businesses::Hired

**Comment by MikeRayUX** [comment]
Alrighty, I think we've made it to a good place in this PR!

The only thing I might want to add is an integration test for "business creates `Forms::Businesses::Hire`"

Other than that it's ready for review.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,12 +5,14 @@
 ### March
 
 * March 20 - Update developer query to not duplicate records with multiple matching specialties #822 @bschrag620
+* March 19 - Added business new hire forms and admin new hire forms #818 @MikeRayUX
```
Comment (joemasilotti) [inline CHANGELOG.md:8]:
```suggestion
* March 21 - Added business new hire forms and admin new hire forms #818 @RomanTurner @MikeRayUX
* March 20 - Update developer query to not duplicate records with multiple matching specialties #822 @bschrag620
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
 * March 18 - Add developer specialties #803
 * March 16 - Don't retry emails for invalid/unsubscribed recipients #798 @zzJZzz
 * March 15 - Fix Preview image doesn't change when select new avatar in profile page #787 @ryy
 * March 10 - Automate subscriber cancellation email #795 @sarvaiyanidhi
 * March 8 - Add a tooltip to replied and email headers on the admin conversations dashboard #796 @franlocus
 * March 7 - Automate subscriber welcome email #794 @duncantmiller
+* March 3 - Added business new hire forms and admin new hire forms #775 @RomanTurner
```
Comment (joemasilotti) [inline CHANGELOG.md:15]:
```suggestion
```

File: app/components/nav_bar/user_component.html.erb [hunk 3]
```diff
@@ -11,24 +11,33 @@
                 <%= inline_svg_tag "logo.svg", class: "h-10 w-auto", alt: "RailsDevs logo", title: "RailsDevs logo" %>
               <% end %>
             </div>
-
```
Comment (joemasilotti) [inline app/components/nav_bar/user_component.html.erb:14]: Can you please add back in all of these deleted blank lines?

File: app/components/nav_bar/user_component.html.erb [hunk 4]
```diff
@@ -11,24 +11,33 @@
...
+                    <%= link_to new_forms_businesses_hire_path, class: "hidden md:block text-gray-300 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium" do %>
+                      <span class="absolute -mt-2.5 -ml-6 z-10 uppercase text-indigo-400 tracking-tight font-bold text-xs -rotate-12"><%= t(".new") %></span>
+                      <span><%= t(".hired_a_dev") %></span>
+                    <% end %>
+                  <% end %>
+                <% elsif business? %>
+                  <%= link_to new_forms_businesses_hire_path, class: "hidden md:block text-gray-300 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium" do %>
+                    <span class="absolute -mt-2.5 -ml-6 z-10 uppercase text-indigo-400 tracking-tight font-bold text-xs -rotate-12"><%= t(".new") %></span>
+                    <span><%= t(".hired_a_dev") %></span>
+                  <% end %>
```
Comment (joemasilotti) [inline app/components/nav_bar/user_component.html.erb:36]: I think this logic can be simplified now.

We are showing both CTAs if a user has both a business and developer profile. So we should be able to check `if developer?` and then `if business?`. Removing the `elsif` should mean not needing the code for the button twice.

Same goes for the mobile portion down below.

File: app/views/admin/forms/businesses/hires/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<%= render Admin::ContainerComponent.new do %>
+  <%= render Admin::ContentComponent.new do %>
+    <%= render Admin::DescriptionListComponent.new(t(".title")) do |list| %>
+      <% list.with_description "Full name" do %>
```
Comment (joemasilotti) [inline app/views/admin/forms/businesses/hires/show.html.erb:4]: i18n, please.

File: app/views/forms/businesses/hires/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<%= form_with model: @form, url: {controller: "forms/businesses/hires", action: "create"} do |form| %>
```
Comment (joemasilotti) [inline app/views/forms/businesses/hires/new.html.erb:1]: I think we can do `form_with model: [:forms, :businesses, :hires, @form]` with the current name spacing.

File: test/fixtures/forms/businesses/hires.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+one:
+  business_id: <%= ActiveRecord::FixtureSet.identify(:one) %>
+  billing_address: 123 sample business st
```
Comment (joemasilotti) [inline test/fixtures/forms/businesses/hires.yml:3]:
```suggestion
  billing_address: 123 Sample Business St.
```

File: test/integration/admin/forms/businesses/hires_test.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
+require "test_helper"
+
+class Admin::Forms::Businesses::HiresTest < ActionDispatch::IntegrationTest
```
Comment (joemasilotti) [inline test/integration/admin/forms/businesses/hires_test.rb:3]: I'm not super interested in testing that the content is on the screen. Can you please update this test to mimic full CRUD actions like `test/integration/admin/specialties_test.rb`?

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking great! I made a few inline requests but outside of the test they are mostly personal code style updates.

However, I think we need to talk about the namespacing again. :)

I love the namespacing that `Forms` creates but I'm thinking it might actually be the wrong abstraction.

There's nothing in the `Forms` namespace that is shared right now. Or even helps distinguish it from other areas. I think what made it clear was how well you named the `<h1>` on the admin screen: "Hiring invoice requests".

I'd love to see this renamed to `Businesses::HiringInvoiceRequest` (and then `Developers::CelebrationPackageRequest` in the future PR) because that is what we are actually calling them in real life.

I understand that that is ANOTHER refactor for you to go through. And probably pretty gnarly as well. 100% OK if you are over this and want me to take over. I appreciate all the work you've put into this PR already! There's a lot of moving parts so getting the naming right will go a long way in making it more intuitive to change in the future.

Let me know what you think and how you'd like to proceed. I also have an additional $25 that a developer asked to pass on to a future issue that I'd love to send your way.

**Comment by joemasilotti** [comment]
Sweet! Thanks @MikeRayUX. Looking forward to your next revision. :)

**Comment by joemasilotti** [comment]
I know there's a lot of changes to be made on this one. Please @ me when everything is ready for review again!

File: app/components/nav_bar/user_component.html.erb (hunk 4 above)
Comment (MikeRayUX) [reply app/components/nav_bar/user_component.html.erb:36]: currently using `hired?` here:

https://github.com/joemasilotti/railsdevs.com/blob/22f7bd53fccd36d6e505bcd69f73a785bdbef446/app/components/nav_bar/user_component.html.erb#L20

Do you prefer `developer?` here instead?

```ruby
# components/nav_bar/user_component.rb

def hired?
  user.developer&.persisted?
end

def developer?
  user.developer&.persisted?
end
```

File: app/components/nav_bar/user_component.html.erb (hunk 3 above)
Comment (MikeRayUX) [reply app/components/nav_bar/user_component.html.erb]: fixed in [f09d2c3](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/f09d2c3ec099818723bd5a229087f5d88b5340b5)

File: test/integration/admin/forms/businesses/hires_test.rb (hunk 8 above)
Comment (MikeRayUX) [reply test/integration/admin/forms/businesses/hires_test.rb:3]: fixed in [f09d2c3](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/f09d2c3ec099818723bd5a229087f5d88b5340b5)

File: db/schema.rb [hunk 9]
```diff
@@ -348,8 +361,8 @@
     t.string "processor_plan", null: false
     t.integer "quantity", default: 1, null: false
     t.string "status", null: false
-    t.datetime "trial_ends_at", precision: nil
-    t.datetime "ends_at", precision: nil
+    t.datetime "trial_ends_at"
+    t.datetime "ends_at"
```
[the remark this answers is not in the document]
Comment (MikeRayUX) [reply db/schema.rb]: reverted in fd19087

File: app/components/nav_bar/user_component.html.erb (hunk 4 above)
Comment (MikeRayUX) [reply app/components/nav_bar/user_component.html.erb]: simplified logic in [0381715](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/03817158013c41b57ea3f1630366320806f3bcb7)

used `developer?` here but can revert to `hired?` if you prefer

Comment (joemasilotti) [reply app/components/nav_bar/user_component.html.erb]: Ah, we need to use the logic in `hired?` because we don't want it to show up on `/developers/new`. Same goes for the business!

I'm down for renaming the method to `developer?` and using that throughout - same for `business?`.

File: test/integration/admin/forms/businesses/hires_test.rb (hunk 8 above)
Comment (MikeRayUX) [reply test/integration/admin/forms/businesses/hires_test.rb:3]: Do you think testing CRUD actions might fit better in a separate test: `Businesses::HiringInvoiceRequestTest`

Since this test just checks for admin content, we could just remove it in favor of above?

Comment (joemasilotti) [reply test/integration/admin/forms/businesses/hires_test.rb:3]: I think we can remove the admin content assertions. But I do have a preference towards integration/feature tests over request tests. Even if the only difference is the name of the file!

Comment (MikeRayUX) [reply test/integration/admin/forms/businesses/hires_test.rb]: changed in [127ce34](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/127ce34565ec1111de4bef3095c1e7caee6c9c52)

Just in case, Updated naming for `Businesses::HiringInvoice` still incoming in refactor.

File: app/views/forms/businesses/hires/new.html.erb (hunk 6 above)
Comment (MikeRayUX) [reply app/views/forms/businesses/hires/new.html.erb:1]: Was able to use `<%= form_with model: @form do |form| %>` here (in upcoming refactor).

File: db/migrate/20230320214634_create_forms_businesses_hires.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+class CreateFormsBusinessesHires < ActiveRecord::Migration[7.0]
+  def change
+    create_table :forms_businesses_hires do |t|
+      t.belongs_to :business, null: false, foreign_key: true
+      t.text :billing_address, null: false
+      t.string :developer_name, null: false
+      t.string :position, null: false
```
[the remark this answers is not in the document]
Comment (MikeRayUX) [reply db/migrate/20230320214634_create_forms_businesses_hires.rb]: fixed in [ae494a8](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/ae494a8c430fca9b6ef3e9e9319491c56b7ba5ea) in the new migration for business_hiring_invoice_requests

https://github.com/joemasilotti/railsdevs.com/blob/ae494a8c430fca9b6ef3e9e9319491c56b7ba5ea/db/migrate/20230321232519_create_businesses_hiring_invoice_requests.rb#L4-L7

File: db/schema.rb [hunk 11]
```diff
@@ -311,7 +324,7 @@
     t.string "processor_id"
     t.boolean "default"
     t.jsonb "data"
-    t.datetime "deleted_at", precision: nil
+    t.datetime "deleted_at"
```
[the remark this answers is not in the document]
Comment (MikeRayUX) [reply db/schema.rb]: reverted in [84a65d3](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/84a65d34353cb3af6caf693a33db5160972cd2ab)

File: db/schema.rb [hunk 12]
```diff
@@ -10,9 +10,8 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.0].define(version: 2023_03_19_212906) do
+ActiveRecord::Schema[7.0].define(version: 2023_03_20_214634) do
   # These are extensions that must be enabled in order to support this database
-  enable_extension "pg_stat_statements"
```
[the remark this answers is not in the document]
Comment (MikeRayUX) [reply db/schema.rb:15]: reverted in [84a65d3](https://github.com/joemasilotti/railsdevs.com/pull/818/commits/84a65d34353cb3af6caf693a33db5160972cd2ab)

**Comment by MikeRayUX** [comment]
@joemasilotti Ready for review!

File: app/components/nav_bar/user_component.html.erb [hunk 13]
```diff
@@ -18,17 +18,22 @@
                 <% links.each do |link| %>
                   <%= helpers.link_to_active link.title, link.path, active_class: "bg-gray-900 text-white", class: "text-gray-300 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium" %>
                 <% end %>
-
```
Comment (joemasilotti) [inline app/components/nav_bar/user_component.html.erb:21]: I think this comment slipped through the cracks. Can you please add these blank lines back in again?

File: test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+module Businesses
+  module HiringInvoiceRequests
+    class Notification::NotificationsTest < ActiveSupport::TestCase
+      include ActionMailer::TestHelper
```
Comment (joemasilotti) [inline test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb:4]:
```suggestion
```suggestion
require "test_helper"

module Businesses
  module HiringInvoiceRequests
    class NotificationsTest < ActiveSupport::TestCase
```

File: test/integration/businesses/hiring_invoice_request_test.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        billing_address: "123 Main St\nNew York, NY 10001",
+        developer_name: "John Doe",
+        position: "Rails Developer",
+        start_date: Date.today,
+        annual_salary: 80_000,
+        employment_type: :full_time_employment,
+        feedback: "John is the perfect candidate!"
+      }
+    }
+  end
```
Comment (joemasilotti) [inline test/integration/businesses/hiring_invoice_request_test.rb:36]: We can get rid of this private method if we include `Admin::Businesses::HiringInvoiceRequestsHelper`.

File: test/support/helpers/businesses/hiring_invoice_requests_helper.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
+module Admin
+  module Businesses
+    module HiringInvoiceRequestsHelper
```
Comment (joemasilotti) [inline test/support/helpers/businesses/hiring_invoice_requests_helper.rb:3]: We can remove the `Admin` module/namespace here.

File: test/fixtures/businesses/hiring_invoice_requests.yml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+one:
+  business_id: <%= ActiveRecord::FixtureSet.identify(:one) %>
```
Comment (joemasilotti) [inline test/fixtures/businesses/hiring_invoice_requests.yml:2]:
```suggestion
  business: businesses(:one)
```

File: test/fixtures/businesses/hiring_invoice_requests.yml [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+  billing_address: 123 Sample Business St.
+  developer_name: John Doe
+  position: Rails Devleoper
+  start_date: <%= Date.current.tomorrow %>
```
Comment (joemasilotti) [inline test/fixtures/businesses/hiring_invoice_requests.yml:6]:
```suggestion
  start_date: <%= Date.current + 7.days %>
```

It's a tiny chance but possible to cross over midnight during a test run.

File: db/schema.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (joemasilotti) [reply db/schema.rb:367]: Looks like these were added back in again.

File: db/migrate/20230321232519_create_businesses_hiring_invoice_requests.rb [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+  def change
+    create_table :businesses_hiring_invoice_requests do |t|
+      t.belongs_to :business, foreign_key: true
+      t.text :billing_address
+      t.string :developer_name
+      t.string :position
+      t.date :start_date
+      t.integer :annual_salary
+      t.integer :employment_type
+      t.text :feedback
```
Comment (joemasilotti) [inline db/migrate/20230321232519_create_businesses_hiring_invoice_requests.rb:11]: Every column that gets validated for presence should have a `null: false` added here.

File: config/routes.rb [hunk 20]
```diff
@@ -38,6 +38,10 @@
       resources :forms, only: [:new, :create], module: :hired
     end
 
+    namespace :businesses do
+      resources :hiring_invoice_requests, only: [:new, :create]
+    end
+
```
Comment (joemasilotti) [inline config/routes.rb:44]: Let's move this to line 17 so it goes "right after" `resources businesses`.

File: config/locales/en.yml [hunk 21]
```diff
@@ -444,7 +472,7 @@ en:
       not_interested: Not interested in work right now
       not_interested_tag: Not interested
       open: Open, but not actively looking
-      open_tag: Open to offers
+      open_tag: Open
```
Comment (joemasilotti) [inline config/locales/en.yml:475]: Please revert this change.

File: config/locales/en.yml [hunk 22]
```diff
@@ -235,6 +251,18 @@ en:
...
+      create:
+        success: Thank you! Joe will reach out to you soon.
+      new:
+        billing_address: Billing address
+        description: Congrats on your new hire!
+        developer_name: Developer's name
+        feedback: Feedback (optional)
+        feedback_help: Please let us know how we can improve RailsDevs or anything you'd like to share.
+        requesting: Requesting...
+        submit: Submit
```
Comment (joemasilotti) [inline config/locales/en.yml:264]:
```suggestion
        request: Request
```

To match the developer form. Make sure to change view, too.

File: app/notifications/admin/businesses/hiring_invoice_request_notification.rb [hunk 23]
```diff
@@ -0,0 +1,22 @@
+module Admin
+  module Businesses
+    class HiringInvoiceRequestNotification < ApplicationNotification
+      deliver_by :database
+      deliver_by :email, mailer: "AdminMailer", method: :businesses_hiring_invoice_request
+
+      param :form
+
+      def title
+        t("notifications.admin.new_hire_form_notification.title", business: form.business.name)
```
Comment (joemasilotti) [inline app/notifications/admin/businesses/hiring_invoice_request_notification.rb:10]: I think this key should match the class name.

File: config/locales/en.yml [hunk 24]
```diff
@@ -541,10 +569,12 @@ en:
       admin: Admin
       billing: Billing
       blocked_conversations: Blocked conversations
+      businesses_hiring_invoice_requests: Business forms
```
Comment (joemasilotti) [inline config/locales/en.yml:572]:
```suggestion
      businesses_hiring_invoice_requests: Hiring invoice requests
```

File: app/views/admin/businesses/hiring_invoice_requests/show.html.erb [hunk 25]
```diff
@@ -0,0 +1,17 @@
+<%= render Admin::ContainerComponent.new do %>
+  <%= render Admin::ContentComponent.new do %>
+    <%= render Admin::DescriptionListComponent.new(t(".title")) do |list| %>
+      <% list.with_description "Full name" do %>
```
Comment (joemasilotti) [inline app/views/admin/businesses/hiring_invoice_requests/show.html.erb:4]: Please move this string to the locale file.

File: app/models/concerns/admin/businesses/hiring_invoice_requests/notifications.rb [hunk 26]
```diff
@@ -0,0 +1,18 @@
+module Admin
```
Comment (joemasilotti) [inline app/models/concerns/admin/businesses/hiring_invoice_requests/notifications.rb:1]: I don't think this should be in the Admin namespace. Yes, we are sending a notification to an admin but that's more implementation details in my opinion.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking great! No major feedback or product changes - it's all working correctly.

The 15 comments are mostly code style and copy tweaks. Let me know if you have any questions.

I think we should be able to merge this soon and then move on to the refactors!

File: db/schema.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeRayUX) [reply db/schema.rb:367]: Think I misunderstood here. Should `precision: nil` exist on these lines?

[the remark this answers is not in the document]
Comment (joemasilotti) [reply db/schema.rb:367]: Honestly, I have no idea :)

But `main` is showing them so let's not make any changes.

File: app/components/nav_bar/user_component.html.erb (hunk 13 above)
Comment (MikeRayUX) [reply app/components/nav_bar/user_component.html.erb:21]: spaces restored in bbec47c455f54437baa20c3e7b1807d326428b9c

File: app/views/admin/businesses/hiring_invoice_requests/show.html.erb (hunk 25 above)
Comment (MikeRayUX) [reply app/views/admin/businesses/hiring_invoice_requests/show.html.erb]: added in e482d5cf1d386557203904931dfe19ffc0356277

File: config/locales/en.yml (hunk 21 above)
Comment (MikeRayUX) [reply config/locales/en.yml]: reverted in f786e1cef45f31d23935218e28516a5d4c236a73

File: app/models/concerns/admin/businesses/hiring_invoice_requests/notifications.rb (hunk 26 above)
Comment (MikeRayUX) [reply app/models/concerns/admin/businesses/hiring_invoice_requests/notifications.rb:1]: changed in e648e383ab02f21a0dbe2e46495bbf7df2df2a3b

File: app/notifications/admin/businesses/hiring_invoice_request_notification.rb (hunk 23 above)
Comment (MikeRayUX) [reply app/notifications/admin/businesses/hiring_invoice_request_notification.rb]: fixed in f11aa8c4682b4abe64e688ec622e83002b1177dc

File: config/routes.rb (hunk 20 above)
Comment (MikeRayUX) [reply config/routes.rb]: moved in efa30812efc2df53eda3669f32ed8a7c01127f7b

File: test/integration/businesses/hiring_invoice_request_test.rb (hunk 15 above)
Comment (MikeRayUX) [reply test/integration/businesses/hiring_invoice_request_test.rb]: changed in e8b7ba8fc79fade9d0d085c8a3a7e9004246f433

File: test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb (hunk 14 above)
Comment (MikeRayUX) [reply test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb:6]: added in d630a6dbcf93efbead42ba91758d5e614891d55a

File: test/support/helpers/businesses/hiring_invoice_requests_helper.rb (hunk 16 above)
Comment (MikeRayUX) [reply test/support/helpers/businesses/hiring_invoice_requests_helper.rb]: module removed in 93591ef79cae96bd3aaba88b5089f04867120fec

File: db/migrate/20230321232519_create_businesses_hiring_invoice_requests.rb (hunk 19 above)
Comment (MikeRayUX) [reply db/migrate/20230321232519_create_businesses_hiring_invoice_requests.rb]: fixed in ae362edf4025f53688e0d722ab91204646b0d42b

File: test/fixtures/businesses/hiring_invoice_requests.yml (hunk 17 above)
Comment (MikeRayUX) [inline test/fixtures/businesses/hiring_invoice_requests.yml:2]: Was unable to get `business: businesses(:one)` or `business_id: businesses(:one).id`  to work here.

Realized `null:false` on the migration affects this.

Thinking `business_id: <%= ActiveRecord::FixtureSet.identify(:one) %>` is the way to go here.

**Comment by MikeRayUX** [comment]
@joemasilotti Ready for review!

**Comment by joemasilotti** [comment]
Can you enable commits on your branch by maintainers? I have a few tiny tweaks I want to make before merging. But this is like 99% of the way there!

**Comment by MikeRayUX** [comment]
"Allow edits by maintainers" box was already checked in the sidebar of the PR.

Is there something else I need to enable somewhere?

**Comment by joemasilotti** [comment]
Weird! No, that should be it.

I'll just add my comments in a review real quick.

**Comment by MikeRayUX** [comment]
>I'll just add my comments in a review real quick.

Ok, ready to implement any changes.

File: test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+
+      test "sends a notification to the admins" do
+        hire_form = Businesses::HiringInvoiceRequest.new(form_attributes)
+        assert_sends_notification Admin::Businesses::HiringInvoiceRequestNotification, to: users(:admin) do
+          assert hire_form.save_and_notify
+        end
+      end
+
+      test "invalid records don't send notifications" do
+        hire_form = Businesses::HiringInvoiceRequest.new
```
Comment (joemasilotti) [inline test/models/concerns/businesses/hiring_invoice_requests/notifications_test.rb:18]:
```suggestion
    class NotificationsTest < ActiveSupport::TestCase
      include ActionMailer::TestHelper
      include HiringInvoiceRequestsHelper
      include NotificationsHelper

      test "sends a notification to the admins" do
        hire_form = HiringInvoiceRequest.new(form_attributes)
        assert_sends_notification Admin::Businesses::HiringInvoiceRequestNotification, to: users(:admin) do
          assert hire_form.save_and_notify
        end
      end

      test "invalid records don't send notifications" do
        hire_form = HiringInvoiceRequest.new
```

File: db/schema.rb [hunk 28]
```diff
@@ -349,8 +363,8 @@
     t.string "processor_plan", null: false
     t.integer "quantity", default: 1, null: false
     t.string "status", null: false
-    t.datetime "trial_ends_at", precision: nil
-    t.datetime "ends_at", precision: nil
+    t.datetime "trial_ends_at"
+    t.datetime "ends_at"
```
Comment (joemasilotti) [inline db/schema.rb:367]:
```suggestion
    t.datetime "trial_ends_at", precision: nil
    t.datetime "ends_at", precision: nil
```

File: test/fixtures/businesses/hiring_invoice_requests.yml [hunk 29]
```diff
@@ -0,0 +1,9 @@
+one:
```
Comment (joemasilotti) [inline test/fixtures/businesses/hiring_invoice_requests.yml:1]: We can delete this file as it isn't used anywhere!

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Last three changes and we are GOOD TO MERGE! 🎉

**Review by joemasilotti — APPROVED** [review]

**Comment by joemasilotti** [comment]
Merged! Thanks for taking this over the finish line @MikeRayUX. And thank you @RomanTurner for kicking this off an getting us some solid foundation.

Can both of you please [email me](mailto:joe@railsdevs.com) how you'd like to receive your bounty? I can do GitHub Sponsorships, Venmo, Zelle, or PayPal.

And @MikeRayUX, whenever you have a minute can you please document the next steps for the refactor? Feel free to assign yourself if you want to take on the work.

**Comment by MikeRayUX** [comment]
Thanks again @RomanTurner for your work in this PR!

It was a pleasure working with you on this Joe!

Happy to file an issue and dive in on the next refactor for `Hired::Form` -> `Developers::CelebrationPackageRequest`

Will get on that first thing in the morning.
